### PR TITLE
mutation/mutation_compactor: validate range tombstone change before it is moved

### PR DESCRIPTION
**Pull request by denesb**
e2c9cdb576c2186719ce7181dd6619f1994e262e moved the validation of the range tombstone change to the place where it is actually consumed, so we don't attempt to pass purged or discarded range tombstones to the validator. In doing so however, the validate pass was moved after the consume call, which moves the range tombstone change, the validator having been passed a moved-from range tombstone. Fix this by moving he validation to before the consume call.

Refs: #12575

### CODE REVIEW

File: mutation/mutation_compactor.hh [hunk 1]
```diff
@@ -197,8 +197,8 @@ private:
         if (_current_emitted_tombstone || (rtc.tombstone() && !can_purge)) {
             partition_is_not_empty(consumer);
             _current_emitted_tombstone = rtc.tombstone();
-            consumer_stop = consumer.consume(std::move(rtc));
             _validator(mutation_fragment_v2::kind::range_tombstone_change, rtc.position(), rtc.tombstone());
+            consumer_stop = consumer.consume(std::move(rtc));
```
Comment (avikivity) [inline mutation/mutation_compactor.hh:201]: Please add a unit test that fails before the fix.

**Comment by denesb** [comment]
v2:
* Add unit test reproducing the issue.

The issue is benign currently, because the validator is configured to not check the clustering key of the validated range tombstone change, which is the only thing destroyed by the earlier move.

**Comment by avikivity** [comment]
Please follow up with a test that feeds badly ordered range tombstones and verifies the correct error is reported. Such a test would have caught this bug, and we also need to make sure the validator isn't regressed later.

**Comment by denesb** [comment]
I did and with the help of the new test, I discovered that https://github.com/scylladb/scylladb/commit/e2c9cdb576c2186719ce7181dd6619f1994e262e in fact broke the design goal of the validator validating the input stream of the compactor (not its output stream).

PR here: https://github.com/scylladb/scylladb/pull/13817

**Comment by denesb** [comment]
> Do we need to backport this? (where's the bot labeling it?)

No need to backport, the bug is benign.
The bot doesn't label pull requests, there was no issue opened for this.

**Pull request merged**